src/ripple/conditions/impl/utils.h

```diff
@@ -31,338 +33,195 @@
+inline
+bool
+isContextSpecific(Preamble const& p)
 {
-    std::stringstream ss;
-    ss << std::dec << value;
-    return ss.str();
+    return (p.type & 0xC0) == 0x80;
 }
 
-// ISO/IEC 8825/7 or ITU-T X.696: Octet Encoding Rules
-// FIXME: This assumes a little-endian architecture!
-namespace oer
+inline
+bool
+isPrivate(Preamble const& p)
 {
+    return (p.type & 0xC0) == 0xC0;
+}
 
-// Simple conversion: write integer as big-endian byte stream
-// This needs to be improved and optimized:
-template <class Integer, class OutputIt>
-void
-encode_integer(Integer value, OutputIt out)
+inline
+Preamble
+parsePreamble(Slice& s, std::error_code& ec)
 {
-    static_assert (
-            std::is_same<Integer, std::uint8_t>::value ||
-            std::is_same<Integer, std::uint16_t>::value ||
-            std::is_same<Integer, std::uint32_t>::value ||
-            std::is_same<Integer, std::uint64_t>::value,
-        "encode_integer accepts only std::uint{8,16,32,64}_t");
+    Preamble p;
 
-    std::size_t n = sizeof(Integer);
-
-    while(n--)
+    if (s.size() < 2)
     {
-        *out++ = static_cast<std::uint8_t>(
-            (value >> (n * 8)) & 0xFF);
+        ec = error::generic;
+        return p;
     }
-}
-
-// Simple conversion: big-endian byte stream to integer
-template <class Integer, class InputIt>
-std::pair<InputIt, Integer>
-decode_integer(InputIt begin, InputIt end)
-{
-    static_assert (
-            std::is_same<Integer, std::uint8_t>::value ||
-            std::is_same<Integer, std::uint16_t>::value ||
-            std::is_same<Integer, std::uint32_t>::value ||
-            std::is_same<Integer, std::uint64_t>::value,
-        "decode_integer accepts only std::uint{8,16,32,64}_t");
-
-    std::size_t size = std::distance (begin, end);
 
-    if (size < sizeof(Integer))
-        Throw<std::length_error>("short integer: " + std::to_string(size));
+    p.type = s[0] & 0xE0;
+    p.tag = s[0] & 0x1F;
 
-    Integer res = 0;
+    s += 1;
 
-    for (std::size_t i = 0; i < sizeof(Integer); ++i)
-        res = (res << 8) | *begin++;
+    if (p.tag == 0x1F)
+    { // Long tag form:
+        p.tag = 0;
 
-    return { begin, res };
-}
+        if (s[0] == 0x80)
+        {
+            ec = error::generic;
+            return p;
+        }
 
-template <class OutputIt>
-inline
-OutputIt
-encode_length (std::size_t len, OutputIt out)
-{
-    if (len <= 0x7F)
-    {
-        *out++ = static_cast<std::uint8_t>(len & 0x7F);
-        return out;
-    }
+        while (!s.empty() && (s[0] & 0x80))
+        {
+            p.tag = p.tag * 128 + (s[0] & 0x7F);
+            s += 1;
+        }
 
-    // Decide how many bytes we need:
-    if (len <= 0xFFFF)
-    {
-        *out++ = 0x82;
-        *out++ = static_cast<std::uint8_t>((len >> 8) & 0xFF);
-        *out++ = static_cast<std::uint8_t>(len & 0xFF);
-        return out;
-    }
+        if (s.empty())
+        {
+            ec = error::generic;
+            return p;
+        }
```

Comment:
Testing s for empty should be moved inside the long tag form as it can never be empty otherwise.